File: Problems/Algorithms/23. Merge K Sorted Lists/MergeKSortedLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<ListNode>(new Comparator<ListNode>() {
            @Override
            public int compare(ListNode node1, ListNode node2){
                if (node1.val < node2.val) {
                    return -1;
                } else if (node1.val == node2.val) {
                    return 0;
                } else {
                    return 1;
                }
            }
        });
        
        for (ListNode node : lists) {
            if (node != null) {
                pq.offer(node);
            }
        }
        
        ListNode dummy = new ListNode(0);
        ListNode node = dummy;
        while (!pq.isEmpty()) {
            ListNode curr = pq.poll();
            node.next = curr;
            node = node.next;
            
            curr = curr.next;
            if (curr != null) {
                pq.offer(curr);
            }
        }
        
        return dummy.next;
    }
}

File: Problems/Algorithms/23. Merge K Sorted Lists/MergeKSortedLists.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun mergeKLists(lists: Array<ListNode?>): ListNode? {
        val pq: PriorityQueue<Pair<Int, ListNode?>> = PriorityQueue<Pair<Int, ListNode?>> { a, b -> a.first - b.first }
        for (list in lists) {
            if (list != null) {
                pq.offer(Pair(list.`val`!!, list))
            }
        }
        
        val dummy = ListNode(0)
        var node = dummy
        while (!pq.isEmpty()) {
            val curr = pq.poll()
            val value = curr.first
            var list = curr.second
            
            node.next = ListNode(value)
            node = node?.next
            
            list = list?.next
            if (list != null) {
                pq.offer(Pair(list.`val`!!, list))
            }
        }
        
        return dummy.next
    }
}

